spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/EmbeddedHeadersMessageConverter.java

```diff
@@ -93,38 +95,52 @@ public static String decodeExceptionMessage(Message<?> requestMessage) {
 	 * have been promoted back to actual headers. The new payload is now the original
 	 * payload.
 	 *
-	 * @param message the message to extract headers
-	 * @param copyRequestHeaders boolean value to specify if original headers should be
-	 * copied
+	 * @param message            the message to extract headers
+	 * @param copyRequestHeaders boolean value to specify if the request headers should be
+	 *                           copied
 	 */
 	public MessageValues extractHeaders(Message<byte[]> message, boolean copyRequestHeaders) throws Exception {
-		byte[] bytes = message.getPayload();
-		ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+		return extractHeaders(message.getPayload(), copyRequestHeaders, message.getHeaders());
+	}
+
+	/**
+	 * Return a message where headers, that were originally embedded into the payload,
+	 * have been promoted back to actual headers. The new payload is now the original
+	 * payload.
+	 *
+	 * @param payload            the message payload
+	 * @param copyRequestHeaders boolean value to specify if the request headers should be copied
+	 * @param requestHeaders     the request headers to copy
+	 * @return the message with extracted headers
+	 * @throws Exception
+	 */
+	public MessageValues extractHeaders(byte[] payload, boolean copyRequestHeaders, MessageHeaders requestHeaders) throws Exception {
```

Comment:
yeah, in that case, I guess we need to have both the methods as `public` as `extractHeaders(payload)` is used by the Kafka binder DLQ forwarding mechanism while the `SendingHandler` requires the other one.